### ISSUE TITLE
Integrate portions of braille extender into NVDA core

### ISSUE THREAD

**Issue by leonardder** [Andre9642/BrailleExtender#31]
I believe braille extender has primarily been developed to overcome shortcomings in NVDA core with regard to braille support. As a member of the NVDA Council who uses braille on a daily basis and lives in a country where braille support is important, I believe that some of its features really belong into the core of a screen reader.

Therefore, I think this issue might be a good platform to start discussing the what, the why and the how about issues that should be ported or integrated into the core of NVDA. We can also have this discussion on the NVDA development list or the add-ons list, but I"m afraid it will make the discussion too scattered and cluttered.

I believe the following Braille Extender features deserve to have a place in NVDA's core base:

* Do not interrupt speech while scrolling braille: nvaccess/nvda#2124. I'm currently working on this
* Font attributes shown with dots 7 and 8 or either one of them: nvaccess/nvda#7097. This issue relates to nvaccess/nvda#1885, nvaccess/nvda#3022 and nvaccess/nvda#6868
* Allow reporting the character you're routing to with cursor routing commands: I don't think there's an NVDA issue for that yet, but I could be wrong
* automatic review cursor tethering in terminal windows.
* switch between several input/output braille tables. In my opinion, this is covered by nvaccess/nvda#8844, which allows you to add multiple configuration profiles and switch to them with shortcuts.
* display tab signs as spaces
* say the current line during text scrolling

cc @Andre9642 @zstanecic @Adriani90 @josephsl

**Comment by Andre9642** [comment]
@leonardder No problem for me! I'd like to contribute more to NVDA Core.
However, I'm afraid because my code is not very clean. This add-on was created 4 years ago. I learned Python (and I still learn) with that. :) I mainly focused on the user experience.
In any case, I'll make better for pull requests in NVDA Core.
I'd also add the auto scroll feature. What do you think?

@zstanecic gesture maps should be improved because there are some conflicts with gestures declared in NVDA Core especially for Freedom Scientific displays. Also, in my opinion, we should discuss the relevance of a generic gesture map for all displays with braille keyboard.

**Comment by Adriani90** [comment]
< Allow reporting the character you're routing to with cursor routing commands>

There is issue nvaccess/nvda#8079.

 

 

Von: Leonard de Ruijter <notifications@github.com> 
Gesendet: Freitag, 25. Januar 2019 08:10
An: Andre9642/BrailleExtender <BrailleExtender@noreply.github.com>
Cc: Adriani90 <adriani.botez@googlemail.com>; Mention <mention@noreply.github.com>
Betreff: [Andre9642/BrailleExtender] Integrate portions of braille extender into NVDA core (#31)

 

I believe braille extender has primarily been developed to overcome shortcomings in NVDA core with regard to braille support. As a member of the NVDA Council who uses braille on a daily basis and lives in a country where braille support is important, I believe that some of its features really belong into the core of a screen reader.

Therefore, I think this issue might be a good platform to start discussing the what, the why and the how about issues that should be ported or integrated into the core of NVDA. We can also have this discussion on the NVDA development list or the add-ons list, but I"m afraid it will make the discussion too scattered and cluttered.

I believe the following Braille Extender features deserve to have a place in NVDA's core base:

*	Do not interrupt speech while scrolling braille: nvaccess/nvda#2124 <https://github.com/nvaccess/nvda/issues/2124> . I'm currently working on this
*	Font attributes shown with dots 7 and 8 or either one of them: nvaccess/nvda#7097 <https://github.com/nvaccess/nvda/issues/7097> . This issue relates to nvaccess/nvda#1885 <https://github.com/nvaccess/nvda/issues/1885> , nvaccess/nvda#3022 <https://github.com/nvaccess/nvda/issues/3022>  and nvaccess/nvda#6868 <https://github.com/nvaccess/nvda/pull/6868> 
*	Allow reporting the character you're routing to with cursor routing commands: I don't think there's an NVDA issue for that yet, but I could be wrong
*	automatic review cursor tethering in terminal windows.
*	switch between several input/output braille tables. In my opinion, this is covered by nvaccess/nvda#8844 <https://github.com/nvaccess/nvda/pull/8844> , which allows you to add multiple configuration profiles and switch to them with shortcuts.
*	display tab signs as spaces
*	say the current line during text scrolling

cc @Andre9642 <https://github.com/Andre9642>  @zstanecic <https://github.com/zstanecic>  @Adriani90 <https://github.com/Adriani90>  @josephsl <https://github.com/josephsl>

**Comment by Adriani90** [comment]
I get also often asked if there is the possibility to use multiple braille output tables at the same time. You can use two tables at the same time with Braille extender. I think this would also be very helpful, especially for people who are reading texts in multiple languages.

**Comment by Adriani90** [comment]
There are also other features whic would be very useful for braille users, but tthose are not covered in Braille extender. Let's progress with the current features first.

**Comment by Adriani90** [comment]
@leonardder the autoscroll feature would be nice also, especially for braille displays for which the scrolling keys are not working or where the driver is not fully completed (i.e. the Baum Vario 80 and 64 displays. But I guess this feature is only working for braille displays which come with it implemented in their own software core. Right?

**Comment by zstanecic** [comment]
Hi,

We cannot do that, as with the current implementation, we’ll have only problems on imput

**Comment by zstanecic** [comment]
Yes, but for eurobraille i don’t have conflicts

**Comment by Andre9642** [comment]
@leonardder Thanks for your feedback! You are probably right.

However, I think that a survey (trough Google Forms or others) should be interesting. Something like:

Question 1: in your opinion, what features of this add-on should be in NVDA Core?

* [ ] Do not interrupt speech while scrolling braille
* [ ] Font attributes shown with dots 7 and 8 or either one of them
* [ ] Allow reporting the character you're routing to with cursor routing commands
* [ ] Automatic review cursor tethering in terminal windows.
* [ ] Switch between several input/output braille tables.
* [ ] Display tab signs as spaces
* [ ] Say the current line during text scrolling
* [ ] Turn on or off dots 7 and 8
* [ ] Autoscroll
* [ ] Complete gesture maps (current maps must be improved)
* [ ] Feedback for volume change

Question 2: what other features would you like to see (in NVDA Core or Braille Extender itself)?
* [ ] Add his own input/output braille tables
* [ ] Output braille table based on language detection
* [ ] Font attributes shown with tags.
* [ ] Font attributes shown with with dots 7 or/and 8 blinking
* [ ] One hand mode
* [ ] Etc.
@leonardder, @zstanecic, @Adriani90, @josephsl and others: what do you think?

**Comment by zstanecic** [comment]
Hi to all,
After a long time, i am responding...
Let's create the survey and pass this around to the NVDA community...
the issue will be interested also to @piciok
Btw, what are plans for developing for this add-on.
Can it be made compatible with Python 3?
@Andre9642 
I've added the latest compatibility flags, but it may break...
Best,
Zvonimir

**Comment by Andre9642** [comment]
@zstanecic Thanks for your comment.
Now I'm in vacation mode until september. So I'll be able to resume my GitHub activity. :)
I'll try to make the survey this weekend and I'll broadcast it trough a future version of the add-on.

**Comment by zstanecic** [comment]
<mailto:reply@reply.github.com> @Andre9642/BrailleExtender

Can you make your add-on compatible with py3?

**Comment by Andre9642** [comment]
> Can you make your add-on compatible with py3?
Yes, I'll try to do it. Currently, I haven't had time to test Pylennium builds.
Thanks

**Comment by josephsl** [comment]
Hello everyone,

Thanks for mentioning me a while back.

I'm looking at the history surrounding this proposal and Braille Extender source code so I can better understand what's going on and come up with a strategy to do some clean ups in preparation for making it easier for Andre to propose pull requests. I might be able to bring in experiences of maintaining Windows 10 App Essentials, Enhanced Touch Gestures, and Add-on Updater, in particular Windows 10 App Essentials as it contains many things that were eventually merged into NVDA.

As Andre pointed out, a general cleanup would be nice, and that's what I plan to do with some sort of a pull request within the next few days. This will mostly involve following Flake8's guidance and updating the add-on source code for Python idiom compliance with NVDA Core expectations (namely variable names, indentation and such). After that, I think we should discuss restructuring the add-on source code so it becomes way easier for Andre to propose pull requests. These may require separate issue/pull request pairs.

For reference, I'm using NVDA alpha snapshots, Braille Extender dev build, and a BrailleNote Touch plus for braille work and to verify that changes do work.

Thanks.

**Comment by josephsl** [comment]
Hi,

At least for me, besides the ones posted above, I'm investigating about the following:

* Auto-scroll (there might be uses for this feature)
* Reversing scroll buttons
* Unicode CLDR character description text in braille (needed for reading emoji descriptions among many others).

Thanks.

**Comment by Andre9642** [comment]
Hi @josephsl,
Regarding auto-scroll, I'm currently rewrite the feature. Please see #81.
Your comments are welcome :)
Thanks

**Comment by dangero2000** [comment]
If merging Braille Extender features into the NVDA core is still ongoing, I'd like to suggest merging speech history. This makes programs previously incapable of outputting to braille displays accessible for braille users by sending speech output to the display. It's a feature I find myself using frequently, most notably in games. I feel this feature is important enough to be part of the core.
